### DISCUSSION TITLE
How to regain access in a self hosted instance?

### DISCUSSION THREAD

**Discussion by crazybolillo** [zitadel/zitadel#8629] [Q&A] [answered]
I was playing around with Zitadel and locked myself out. Basically I changed the admin's email and password and once I went in to log in again, Zitadel asked for a verification code that was sent to the new email, however since outgoing email was not configured, I never got the code and now I am unable to log in.

I have access to the CLI and Database. How would I go about unlocking myself? I am just playing around so it is not big deal, I can just delete the containers and start over, but it does make me wonder if there is some "escape hatch" you can use to reset password or gain admin access through a CLI in case something goes wrong.

For example, grafana [lets you reset the admin password from their CLI](https://grafana.com/docs/grafana/latest/cli/#reset-admin-password).

**Comment by hifabienne** [comment]
It is possible by editing the user in the database, as far as I know there should be a table called something like projections.users_x
where you can edit the user, but i am not 100% sure

@adlerhurst @muhlemmer can you help.

**Comment by muhlemmer** [comment, marked as answer]
You can create a [System User](https://zitadel.com/docs/guides/integrate/zitadel-apis/access-zitadel-system-api#system-api-user) with the appropriate roles. (IAM_OWNER or ORG_OWNER etc) and create a JWT.

Then call the user API with the token and update the user: https://zitadel.com/docs/apis/resources/user_service_v2/user-service-update-human-user

You can use [gRPCui](https://github.com/fullstorydev/grpcui) to call the API in a more friendly way.

```
go install github.com/fullstorydev/grpcui/cmd/grpcui@latest
grcpui -plaintext localhost:9000
```

Set `Authorization`: `Bearer <SYSTEM_JWT>` in the Request Metadata in the UI, and switch to the user service.

**Reply by crazybolillo** [reply]
Thanks, I hope I don't ever lock myself out but it comes in handy, and I did not know it was possible to set up system users through the config file.

**Reply by zoechi** [reply]
I also just found myself in this situation.
It looks like `user-service-update-human-user` requires the old password to set a new one.
  
I just set up Zitadel and I had it configured to require an admin password change at first login but my password manager saved the old password instead of the new one.

I'm even still logged in, but I haven't found a way to get my situation fixed without starting the whole configuration over.

**Comment by RD-1-2020** [comment]
Hello, i write solutions for docker install 

1. generate private key (.pem) and pub key (.pub) as in instruction https://zitadel.com/docs/guides/integrate/zitadel-apis/access-zitadel-system-api#system-api-user
2. generate base64 from you pub key(.pub) data
3. Add env in you docker .env file or in docker-compose.yml file (in this json 'systemuser' this username from step 1 manual)
 ```
ZITADEL_SYSTEMAPIUSERS='{"systemuser":{"KeyData":"<YOUR_PUB_KEY_BASE_64_DATA>","Memberships":[{"MemberType":"System","Roles":["SYSTEM_OWNER","IAM_OWNER","ORG_OWNER"]},{"MemberType":"IAM","Roles":"IAM_OWNER","AggregateID":"123456789012345678"},{"MemberType":"Organization","Roles":"ORG_OWNER","AggregateID":"123456789012345678"}]}}'
```
4. With zitadel-tools and private key (.pem) file get JWT `zitadel-tools key2jwt --audience=https://YOU_ZITADEL_HOSTNAME --key=systemuser.pem --issuer=systemuser`
5. With CURL use API

```
curl -L 'https://YOU_ZITADEL_HOSTNAME/management/v1/users/:user_id/_unlock' -H 'Content-Type: application/json' -H 'Accept: application/json' -H 'Authorization: Bearer <YOU JWT FROM 4 STEP>' -d '{}'
```

You can get user_id from requests in network from you browser or you can find id in database.

<img width="2315" height="540" alt="изображение" src="https://github.com/user-attachments/assets/eaf517b7-4cdb-4192-b7fc-c85fb40b452d" />

Find in db:

1. sudo docker compose exec -it zdb bash
2. psql -U root -d zitadel (or use your zitadel user)
3. select * from projections.users13; (your table name may looks like usersXXX)

<img width="1814" height="165" alt="изображение" src="https://github.com/user-attachments/assets/d3fabc17-210a-43b6-b196-4f0a28570d9e" />

**Reply by wanderingMonk88** [reply]
I am in a similar boat. I tried these instructions. 
1. You may want to update step 3 and 4 here to reference user-system-1 to be more aligned with the documentation linked in step 1.
2. Do you want to update the last step to add optional detail to connect to projections.login_names3_users
3. More importantly I still get Errors.Token.Invalid. Do you have any suggestions?

**Reply by RD-1-2020** [reply]
I tomorrow spent 4 hours to unlock my admin, this sooo difficult...)

I try help you)

- Your Zitadel installation is Docker?
- Your json value of ZITADEL_SYSTEMAPIUSERS surrounded in '...'?
- Are you paste the base64(.pub) inside KeyData?
- Are you sure your private key on generate JWT the same private key of generated public key from KeyData? You can check this with openssl and online?

If my questions don't help you:

Maybe you can show me your: 
- Zitadel env with blurred pub key and username 
- Zitadel tools command for generate JWT without JWT and your hostname
- Zitadel container logs with JWT verification errors or warnings

**Reply by wanderingMonk88** [reply]
Thank you so much for the help. 
* Yes the installation is docker
* It is surrounded in '.....'
* Used the output of `cat system-user-1.pub | base64`
* Tried it out on jwt.io and it is valid and verified using the public key

Thanks for asking for those debug messages. I found the following in the logs:
`2025-11-01T18:03:04.429316843Z time="2025-11-01T18:03:04Z" level=warning msg="token verifier repo: decrypt access token" caller="/home/runner/work/zitadel/zitadel/internal/authz/repository/eventsourcing/eventstore/token_verifier.go:303" error="ID=APP-ASdgg Message=invalid token"
2025-11-01T18:03:04.429367554Z time="2025-11-01T18:03:04Z" level=warning msg="token verifier repo: verify JWT access token" caller="/home/runner/work/zitadel/zitadel/internal/authz/repository/eventsourcing/eventstore/token_verifier.go:307" error="issuer does not match: Expected: http://odyssey-zitadel.lan, got: system-user-1"
2025-11-01T18:03:04.429381604Z time="2025-11-01T18:03:04Z" level=warning msg="authz: verify access token" caller="/home/runner/work/zitadel/zitadel/internal/api/authz/context.go:114" error="ID=APP-Reb32 Message=invalid token" org_domain= org_id=`

That was confusing.  The decoded payload looks like:
`{
  "iss": "system-user-1",
  "sub": "system-user-1",
  "aud": [
    "http://odyssey-zitadel.lan"
  ],
  "iat": 1762020024,
  "exp": 1762023624
}` on jwt.io

https://github.com/zitadel/zitadel/issues/7937 seemed related, which is when the realization of missing port was the issue. I had to add `:8080` to the audience.

**Reply by RD-1-2020** [reply]
This looks like Zitadel does not read user from env....

Look docker inspect. Maybe you see any anomaly in this output. 

And I try thinking about this problem. For more information you can give env without public key.

**Reply by RD-1-2020** [reply]
For the history, I'd like to record that I encountered an interesting problem.

If there's a time discrepancy between the machine you're generating the JWT on and the Zitadel server, you'll receive an Invalid Token error until you resolve the time discrepancy.

**Comment by wanderingMonk88** [comment]
I want to be clear. Thanks to your help, I was able to get the System APIs working with the addition of port to the audience. Thanks again.

**Comment by IJustDev** [comment]
This is by far the worst recovery I had to do in my entire life. Maybe there should be a simpler mechanism for the following generations...

**Reply by IJustDev** [reply]
So I followed the attached documentation https://zitadel.com/docs/guides/integrate/zitadel-apis/access-zitadel-system-api#system-api-user.

The logs of zitadel told me that I the issuers do not match and it expects something like https://domain instead of `system-user-1`.
Do not fall for it. The issuer remains system-user-1.

In my case the audience was the problem. It is not https://<domain>. it is https://<domain>:<exposedPort>
